ai/prompts/PYRE_ANALYSIS.md

```diff
@@ -0,0 +1,254 @@
+pyre.bin: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found
+```
+→ Use Docker approach instead.
+
+## Top 5 Pyre Commands for Analysis
+
+### 1. Find All References to a Function/Class
+
+```bash
+# Find everywhere a method is called
+pyre query "find_references('graphistry.PlotterBase.nodes')"
+
+# Find everywhere a class is used
+pyre query "find_references('graphistry.Plottable.Plottable')"
+```
+
+**Use Case**: Before refactoring a method, find all call sites to assess impact.
+
+### 2. Get Type Information
+
+```bash
+# Get the type of an expression at a specific location
+pyre query "types('graphistry/PlotterBase.py')"
+
+# Get the type at a specific line and column
+pyre query "type_at_position('graphistry/PlotterBase.py', 100, 10)"
+```
+
+**Use Case**: Understanding complex type flows, verifying inferred types.
+
+### 3. Find All Attributes
+
+```bash
+# Find all attributes of a class
+pyre query "attributes('graphistry.PlotterBase.PlotterBase')"
+
+# Find where an attribute is defined
+pyre query "defines('graphistry.PlotterBase.PlotterBase', '_dataset_id')"
+```
+
+**Use Case**: Attribute tracking, finding all dataset-relevant attributes (as in our ID invalidation bug).
+
+### 4. Call Graph Analysis
+
+```bash
+# Find all callees of a function (what it calls)
+pyre query "callees('graphistry.PlotterBase.nodes')"
+
+# Find all callers of a function (what calls it)
+pyre query "callers('graphistry.PlotterBase.nodes')"
+```
+
+**Use Case**: Understanding data flow, finding indirect modification paths.
+
+### 5. Validate Types
+
+```bash
+# Check types without query mode
+pyre check
+
+# Check with detailed output
+pyre check --output=json
+
+# Check specific file
+pyre check graphistry/PlotterBase.py
+```
+
+**Use Case**: Pre-commit validation, finding type errors.
+
+## Practical Example: Finding Dataset ID Invalidation Issues
+
+**Problem**: Find all methods that modify `_dataset_id` attribute.
+
+### Attempt 1: Pyre Query (Ideal but slow)
+```bash
+docker run --rm -v $(pwd):/workspace -w /workspace python:3.12-slim bash -c "
+    pip install pyre-check && \
+    pyre query \"find_references('graphistry.PlotterBase._dataset_id')\"
+"
+```
+
+**Issue**: May timeout on large codebase (5-10 minutes).
+
+### Attempt 2: AST Script (Practical alternative)
+```python
+# plans/fix_dataset_file_id_invalidation/analyze_attribute_modifications.py
+import ast
+
+class MethodAnalyzer(ast.NodeVisitor):
+    def visit_Assign(self, node):
+        # Track all assignments to _dataset_id
+        if isinstance(target, ast.Attribute) and target.attr == '_dataset_id':
+            # Record method name and line number
+            ...
+```
+
+**Result**: Completes in < 1 second, found all 5 buggy methods.
+
+**Lesson**: For custom analysis patterns, AST scripts are often more practical than pyre queries.
+
+## Troubleshooting
+
+### Pyre Times Out
+
+**Symptom**: Hangs at "Processed 1631 of 1709 functions" for 5+ minutes.
+
+**Solutions**:
```

Comment:
if timeout, turn on verbose/debug mode to idnetify problematic method and see if can solve such as fixing file or ignoring it